content/blog/2022-04-28-chlorophyll/index.md

```diff
@@ -0,0 +1,309 @@
+
+<center>
+<img src="sliding-window.png" alt="alt_text" title="image_tooltip" style="zoom:20%;" />
+</center>
+
+## Interaction Between Steps
+### Iterative Refinement Method
+Since the above four steps are not aware of each other, it may be possible that some optimizations in specific steps are not suitable for others. For example, passing a message from the core $A$ to core $B$, the layout result may require the message to be passed by $C$, but in the previous step, $C$ is not aware of that, so the message buffer is not counted into $C$'s partition space, which may exceed the memory of $C$. The authors propose an iterative refinement method to relax the communication constraints until all data fit in each of the cores.
+
+### Optimization Opportunity Loss
+Since the authors decompose the problem into four subproblems and solve them individually, which may lead to possible optimization opportunity loss:
+1. <u>Partition before superoptimization.</u> For example, A, B, and C are three code segments. A+B may have the lowest communication cost, but B+C may have the lowest computation cost if they are put on the same core. The partition algorithm is not aware of that.
+2. <u>Schedule-oblivious routing strategy.</u> The routing algorithm does not know which core is busy and may still send messages to the busy core.
+3. <u>Scope of the superoptimizer.</u> The superoptimizer leverages a greedy algorithm, which works well for local code segments but is not the best for the whole program.
+
+
+## Performance Evaluation
+### Execution Time
+The authors evaluate their compiler on different benchmarks and compare the performance of different settings (with/without sliding-window superoptimization, with/without partitioning synthesis, and with/without layout synthesis). The results are shown below. But since the author only makes self-comparison, it is hard to say whether the results are good enough.
+<center>
+<img src="perf.png" alt="alt_text" title="image_tooltip" style="zoom:50%;" />
+</center>
+
+They do compare their generated code with handwritten code and give the following statement:
+> "Compared to the experts' implementation, it is **only 65% slower, 70% less energy-efficient and uses 2.2x more cores**. This result confirms that our generated programs are comparable with experts’ not only on small programs but also on a real application."
+
+The authors are somehow proud of this result. Considering Chlorophyll is the first high-level compiler for GA144, if it indeed reduces the programming effort, then even the performance is not good, it is useful for more applications running on GA144.
```

Comment:
> The authors are somehow proud of this result.

😅 That may be true—but to be pedantic about it, the authors didn't ever say anything about how they _felt_ about the results! Just that they are "comparable" with expert solutions, which is somewhat vague. I might recommend avoiding direct statements about what is in the authors' minds.